### PR TITLE
Implement PROVIDES instruction

### PR DESCRIPTION
**Pull request by thekid**
See [this comment on #12](https://github.com/tueftler/doget/issues/12#issuecomment-245221570): The `PROVIDES` instruction declares a trait provides a certain image when included, e.g. "php:latest". Other traits can then happily extend from these in their `FROM` statement.

## Example

```dockerfile
# Enforce corporate base image policy
FROM corporate.example.com/debian:jessie

# Provide "java:openjdk-8-jre" in the easiest way possible
RUN set -x && apt-get update && apt-get install -y \
    openjdk-8-jre-headless \
  && rm -rf /var/lib/apt/lists/*
PROVIDES java:openjdk-8-jre

# This trait inherits from  "java:openjdk-8-jre". Use exact SHA since this is what we
# audited for security, and it will always reference the exact file contents we looked at
USE github.com/neo4j/docker-neo4j-publish/3.0.4/community/Dockerfile:1ad894c…

# Add tini
ENV TINI_VERSION v0.10.0
ADD https://github.com/krallin/tini/releases/download/${TINI_VERSION}/tini /tini
RUN chmod +x /tini

# Overwrite Neo4J's entrypoint 
ENTRYPOINT ["/tini", "--", "/docker-entrypoint.sh"]
```

### CODE REVIEW

**Comment by thekid** [comment]
Hrm, after some thought, we don't need `AS` after all:

```dockerfile
USE github.com/docker-library/php/7.1 AS php:latest

# ...is the same as:
USE github.com/docker-library/php/7.1
PROVIDES php:latest
```

File: use/extension.go [hunk 1]
```diff
@@ -56,27 +56,32 @@ func (s *Statement) Emit(out io.Writer) {
 // Origin parses origin from reference
```
Comment (mikey179) [inline use/extension.go]: Being a bit picky, but the changes in this file are unrelated to the PR, aren't they?

File: command/transform/transformation.go [hunk 2]
```diff
@@ -56,7 +58,7 @@ func (t *Transformation) Run(parser *dockerfile.Parser) error {
 	}
 
 	file.From.Emit(t.Output)
-	return t.write(parser, &file, "")
+	return t.write(parser, &file, "", map[string]bool{file.From.Image: true})
```
Comment (mikey179) [inline command/transform/transformation.go]: I think we should introduce a separate type for the list of provided images within the `provided` package. Will make it more clear what is happening here, and allow to hide some of the specific details of the handling in the `write()` method below.

File: use/extension.go (hunk 1 above)
Comment (thekid) [inline use/extension.go]: These specific changes are boyscout ones, yes.

File: command/transform/transformation.go (hunk 2 above)
Comment (thekid) [inline command/transform/transformation.go]: You mean like this?

```diff
--- a/command/transform/transformation.go
+++ b/command/transform/transformation.go
@@ -18,6 +18,8 @@ type Transformation struct {
        Output io.Writer
 }

+type Provided map[string]bool
+
 func parse(parser *dockerfile.Parser, input string, file *dockerfile.Dockerfile) error {
        if err := parser.ParseFile(input, file); err != nil {
                return err
@@ -58,7 +60,7 @@ func (t *Transformation) Run(parser *dockerfile.Parser) error {
        }

        file.From.Emit(t.Output)
-       return t.write(parser, &file, "", map[string]bool{file.From.Image: true})
+       return t.write(parser, &file, "", Provided{file.From.Image: true})
 }

 func prefix(paths, base string) string {
@@ -74,7 +76,7 @@ func prefix(paths, base string) string {
        return result + segments[len(segments)-1]
 }

-func (t *Transformation) write(parser *dockerfile.Parser, file *dockerfile.Dockerfile, base string, provided map[string]bool) error {
+func (t *Transformation) write(parser *dockerfile.Parser, file *dockerfile.Dockerfile, base string, provided Provided) error {
        for _, statement := range file.Statements {
                switch statement.(type) {
                case *provides.Statement:`
```

Comment (mikey179) [inline command/transform/transformation.go]: Something like this, but with methods that encapsulate what's happening in lines 81-84 and the check in line 118:

```
func  (p *Provides) add(images []string)
func (p *Provides) include(image) bool
```

Comment (thekid) [inline command/transform/transformation.go]: Done!
